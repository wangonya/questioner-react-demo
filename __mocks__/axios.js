let currentId = 2;

module.exports = {
  get: url => {
    return Promise.resolve({
      data: [
        {
          id: 2,
          title: "Another test meetup",
          details: "Yet another test meetup. Hope it works too.",
          creator: 1,
          location: "Some venue somewhere",
          happening_on: "2019-02-27",
          created_on: "2019-02-20",
          tags: "new, test",
          image: "https://ipsum-media.netlify.com/img/05.jpg"
        },
        {
          id: 3,
          title: "Yet another test meetup",
          details: "Yet another test meetup. Hope it works too.",
          creator: 1,
          location: "Some venue somewhere",
          happening_on: "2019-02-28",
          created_on: "2019-02-20",
          tags: "new, test",
          image: "https://ipsum-media.netlify.com/img/06.jpg"
        },
        {
          id: 4,
          title: "Small scale farming: Value chain",
          details:
            "Abstract/Descrincreasingly strict product standards, and the farmersâ€™ aversion to risk.",
          creator: 1,
          location: "Kisumu-Kenya",
          happening_on: "2019-03-29",
          created_on: "2019-03-23",
          tags: "Farming, small scale, market",
          image: "/home/richard/Desktop/cat.jpg"
        }
      ]
    });
  },
  post: url => {
    return Promise.resolve({
      data: {
        task: {
          name: currentId,
          id: currentId++
        }
      }
    });
  }
};
