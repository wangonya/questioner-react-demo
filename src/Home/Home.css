.meetups {
  display: grid;
  grid-template-columns: repeat(auto-fill, minmax(auto, auto));
  grid-gap: 10px;
  grid-auto-rows: minmax(180px, auto);
  grid-auto-flow: dense;
  padding-top: 10px;
}

.home .grid-item {
  padding: 1rem;
  background: linear-gradient(to bottom, rgba(33, 33, 33, 0) 0%, black 100%);
  border-radius: 5px;
  transition: box-shadow 0.3s;
  position: relative;
  overflow: hidden;
}

.home .grid-item > img {
  object-fit: cover;
  width: 100%;
  height: 100%;
  position: absolute;
  z-index: -1;
  top: 0;
  left: 0;
  transition: all 1s ease;
}

.home .grid-item:hover img {
  transform: scale(1.02);
}

.home .main {
  grid-column-end: span 6;
  grid-row-end: span 2;
}

.home .featured {
  grid-column-end: span 3;
  grid-row-end: span 1;
}

.text-block {
  position: absolute;
  bottom: 20px;
  left: 20px;
  color: white;
  padding-left: 20px;
  padding-right: 20px;
}

@media (max-width: 800px) {
  .home .featured {
    grid-column-end: span 6;
    grid-row-end: span 1;
  }
}
