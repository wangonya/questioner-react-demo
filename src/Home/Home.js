import React, { Component } from "react";
import axios from "axios";

import "./Home.css";

import Loader from ".././Loader/Loader.js";

class Home extends Component {
  constructor(props) {
    super(props);

    this.state = {
      meetups: [],
      loading: true,
      error: null
    };
  }

  componentDidMount() {
    axios
      .get("https://questioner2.herokuapp.com/api/v2/meetups/upcoming")
      .then(res => {
        const meetups = res.data.data;

        // update state to trigger re-render
        this.setState({
          meetups,
          loading: false,
          error: null
        });
      })
      .catch(err => {
        // save err in state and re-render
        this.setState({
          loading: false,
          error: err
        });
      });
  }

  renderLoading() {
    return <Loader />;
  }

  renderError() {
    return <div>Uh oh: {this.state.error.message}</div>;
  }

  renderMeetups() {
    if (this.state.error) {
      return this.renderError();
    }

    let main_meetup = this.state.meetups[0];

    return (
      <div className="meetups">
        <div className="grid-item main">
          <img src={main_meetup.image} alt="meetup image" />
          <div className="text-block">
            <a href="">
              <h2>{main_meetup.title}</h2>
            </a>
            <small>{main_meetup.happening_on}</small>
            <br />
            <p>{main_meetup.details}</p>
          </div>
        </div>
        {this.state.meetups.splice(1).map(meetup => (
          <div className="grid-item featured" key={meetup.id}>
            <img src={meetup.image} alt="meetup image" />
            <div className="text-block">
              <a href="">
                <h4>{meetup.title}</h4>
              </a>
              <small>{meetup.happening_on}</small>
            </div>
          </div>
        ))}
      </div>
    );
  }

  render() {
    return (
      <div className="home">
        <h2 className="title">Upcoming Meetups</h2>
        {this.state.loading ? this.renderLoading() : this.renderMeetups()}
      </div>
    );
  }
}

export default Home;
