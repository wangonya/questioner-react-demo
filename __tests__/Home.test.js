import React from "react";
import { shallow } from "enzyme";
import axios from "axios";

import Home from ".././src/Home/Home";

jest.mock("axios");

describe("<Home /> component", () => {
  const getSpy = jest.spyOn(axios, "get");
  const wrapper = shallow(<Home />);

  test("renders without crashing", () => {
    expect(wrapper).toMatchSnapshot();
  });

  test("renders with correct state", () => {
    expect(shallow(<Home />).state("meetups")).toEqual([]);
    expect(shallow(<Home />).state("loading")).toEqual(true);
    expect(shallow(<Home />).state("error")).toEqual(null);
  });

  test("mock fetch a list of meetups from api", () => {
    expect(getSpy).toBeCalled();
    getSpy.mockClear();
  });

  test("renders error from state", () => {
    const errWrapper = shallow(<Home />);
    expect(errWrapper.find("div.err")).toHaveLength(0);
    errWrapper.setState({ error: true, loading: false });
    expect(errWrapper.state("error")).toEqual(true);
    expect(errWrapper.find("div.err")).toHaveLength(1);
  });
});
