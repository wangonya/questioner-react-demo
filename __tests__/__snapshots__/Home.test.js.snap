// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Home /> component renders without crashing 1`] = `ShallowWrapper {}`;
